Ignore IsNotFound error

FengyunPan opened the pull request:
IsNotFound error is fine since that means the object is
deleted already, so we should check err and ignore err
before returning.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44483)
<!-- Reviewable:end -->

FengyunPan commented:
/release-note-none

quinton-hoole commented:
@k8s-bot ok to test

quinton-hoole commented:
This looks good to me.  @marun would you like to do a quick second review, as you've had your head deep in this stuff lately.

quinton-hoole commented:
/lgtm

quinton-hoole commented:
/approve cancel

quinton-hoole commented:
Somehow my LGTM also approved this, which is not what I intended.  Waiting for a second review before approving.

marun commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -176,9 +176,15 @@ func newFederationSyncController(client federationclientset.Interface, adapter f
 			return err
 		},
 		func(client kubeclientset.Interface, obj pkgruntime.Object) error {
+			secret := obj.(*apiv1.Secret)
```
This controller is generic - it's not intended to be just for secrets.  And a secret isn't needed here - that's what namespacedName is for.

FengyunPan replied:
Done

marun commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -176,9 +176,15 @@ func newFederationSyncController(client federationclientset.Interface, adapter f
 			err := adapter.ClusterDelete(client, namespacedName, &metav1.DeleteOptions{OrphanDependents: &orphanDependents})
+			// IsNotFound error is fine since that means the object is deleted already.
+			if errors.IsNotFound(err) {
+				glog.V(4).Infof("Secret %s/%s no longer exists", secret.Namespace, secret.Name)
```
```golang
glog.V(4).Infof("%s %q no longer exists",  strings.ToTitle(adapter.Kind()), namespacedName)
```
The name of the type being federated can be retrieved via ``adapter.Kind()`` (returns lowercase), and namespacedName's ``String`` method results in ``[namespace]/[name]``.

FengyunPan replied:
Nice! Done.

marun reviewed (commented):
Consider updating the federated updater to check the error for ``IsNotFound`` instead of making the change per-type.  The logging isn't really necessary since a NotFound error is safe to ignore in this case.

marun commented:
As per my review comment, why is it preferable to implement controller-specific logic?  Why not simply ignore ``NotFound`` in [FederatedUpdater.UpdateWithOnError](https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/util/federated_updater.go#L96):

```golang
switch op.Type {
...
case OperationTypeDelete:
  err = fu.deleteFunction(clientset, op.Obj)
  if errors.IsNotFound(err) {
    err = nil
  }
}
```

marun commented:
@k8s-bot non-cri e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

marun commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -115,7 +115,7 @@ func newFederationSyncController(client federationclientset.Interface, adapter f
 	}
 
-	// Build delivereres for triggering reconciliations.
+	// Build deliverers for triggering reconciliations.
```
Please separate these unrelated and non-functional changes into a separate commit.

Commit:
Ignore IsNotFound error

IsNotFound error is fine since that means the object is
deleted already, so we should ignore it.

marun commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

marun commented:
/lgtm

madhusudancs commented:
@k8s-bot kops aws e2e test this

madhusudancs commented:
@FengyunPan could you please ping me once all the tests pass here - http://prow.k8s.io/?type=presubmit&repo=kubernetes%2Fkubernetes&pull=44483

FengyunPan commented:
@madhusudancs yes.

FengyunPan commented:
@madhusudancs Hi, the pull-kubernetes-e2e-kubeadm-gce is FAILURE, but I think that is not the result of this patch.

madhusudancs commented:
@FengyunPan kubeadm test is not required, so we can ignore that.

We need federation tests to pass for federation PRs and it failed last time. I am going to run it again. You can monitor the progress at the same URL.

@k8s-bot pull-kubernetes-federation-e2e-gce test this

FengyunPan commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot kops aws e2e test this

marun commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

Results: http://prow.k8s.io/?type=presubmit&pull=44483&job=pull-kubernetes-federation-e2e-gce

madhusudancs commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

madhusudancs commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

madhusudancs commented:
Federation presubmit passed - https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/44483/pull-kubernetes-federation-e2e-gce/2698/

madhusudancs commented:
/lgtm
/approve
